api/src/AppBundle/Service/CourtOrderCreator.php

```diff
@@ -0,0 +1,171 @@
+
+        return $courtOrder;
+    }
+
+    /**
+     * Replace an existing court order with a dulpicate
+     */
+    private function recreateCourtOrder(CourtOrder $courtOrder): CourtOrder
+    {
+        $orderDto = (new CourtOrderDto())
+            ->setCaseNumber($courtOrder->getCaseNumber())
+            ->setType($courtOrder->getType())
+            ->setSupervisionLevel($courtOrder->getSupervisionLevel())
+            ->setOrderDate($courtOrder->getOrderDate());
+        $client = $courtOrder->getClient();
+        $report = $courtOrder->getReports()[0];
+
+        $this->em->remove($courtOrder);
+        $this->em->flush();
+
+        return $this->createCourtOrder($orderDto, $client, $report);
+    }
+
+    private function createCourtOrderDeputy(CourtOrderDeputyDto $deputyDto, CourtOrder $courtOrder, Organisation $organisation): CourtOrderDeputy
+    {
+        $deputy = $this->courtOrderDeputyFactory->create($deputyDto, $courtOrder);
+        $deputy->setOrganisation($organisation);
+
+        $this->em->persist($deputy);
+        $this->em->flush();
+
+        return $deputy;
+    }
+
+    private function updateCourtOrderDeputy(CourtOrderDeputy $deputy, CourtOrderDeputyDto $deputyDto, Organisation $organisation = null): void
+    {
+        $deputy
+            ->setFirstname($deputyDto->getFirstname())
+            ->setSurname($deputyDto->getSurname())
+            ->setEmail($deputyDto->getEmail());
+
+        $addressDto = $deputyDto->getAddress();
+        $deputy->getAddresses()[0]
+            ->setAddressLine1($addressDto->getAddressLine1())
+            ->setAddressLine2($addressDto->getAddressLine2())
+            ->setAddressLine3($addressDto->getAddressLine3())
+            ->setTown($addressDto->getTown())
+            ->setCounty($addressDto->getCounty())
+            ->setPostcode($addressDto->getPostcode())
+            ->setCountry($addressDto->getCountry());
+
+        if (!is_null($organisation)) {
+            $deputy->setOrganisation($organisation);
+        }
+    }
+
+    public function upsertCourtOrder(CourtOrderDto $orderDto, Report $report): CourtOrder
+    {
+        $client = $report->getClient();
+
+        $courtOrder = $this->getCourtOrder($client, $orderDto->getType());
+
+        if (is_null($courtOrder)) {
+            return $this->createCourtOrder($orderDto, $client, $report);
+        } else {
+            $courtOrder->setOrderDate($orderDto->getOrderDate());
+            return $courtOrder;
+        }
+    }
+
+    public function upsertCourtOrderDeputy(CourtOrderDeputyDto $deputyDto, CourtOrder $courtOrder, Organisation $organisation): CourtOrderDeputy
+    {
+        if ($this->courtOrderHasLayDeputy($courtOrder)) {
+            throw new \RuntimeException('Case number already used by lay deputy');
+        }
+
+        if (count($courtOrder->getDeputies()) > 1) {
+            throw new \RuntimeException('Court order has multiple organisations attached');
+        }
```

Comment:
Can you briefly explain this scenario please? :)

Reply:
In the future we'll want to support a court order having multiple deputies, so we can model Sirius properly. To that end, the database relationship is one-to-many. But at the moment we don't actually support that, because CasRec doesn't provide the right data.

This line should never be hit, because our code doesn't currently add two deputies to the same order. It _could_ be hit if there was some data corruption, dodgy code, or someone messed with the database directly. This catches that "something is wrong" state and stops us making assumptions about whether the deputy should be updated or replaced.

It's basically the same as the "Can't determine if deputy has moved with client to new org" error we have for messed-up NamedDeputy records.

Reply:
Thanks